source/agora/common/crypto/Schnorr.d

```diff
@@ -0,0 +1,442 @@
+    time and `O(1)` and `O(1)` signature size.
+    Additionally, since the `c` factor does not depend on EC operation,
+    we can do batch verification, enabling us to speed up verification
+    when verifying large amount of data (e.g. a block).
+
+    See_Also:
+      - https://en.wikipedia.org/wiki/Curve25519
+      - https://en.wikipedia.org/wiki/Schnorr_signature
+      - https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki
+      - https://www.secg.org/sec1-v2.pdf
+
+    TODO:
+      - Compress signature according to SEC1 v2 (Section 2.3)
+      - Audit GDC and LDC generated code
+      - Proper audit
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.crypto.Schnorr;
+
+import agora.common.Data;
+import agora.common.Hash;
+
+import geod24.bitblob;
+import libsodium;
+
+import std.algorithm;
+import std.range;
+
+
+/// Single signature example
+nothrow @nogc unittest
+{
+    Pair kp = Pair.random();
+    auto signature = sign(kp, "Hello world");
+    assert(verify(kp.V, signature, "Hello world"));
+}
+
+/// Multi-signature example
+nothrow @nogc unittest
+{
+    // Setup
+    static immutable string secret = "BOSAGORA for the win";
+    Pair kp1 = Pair.random();
+    Pair kp2 = Pair.random();
+    Pair R1 = Pair.random();
+    Pair R2 = Pair.random();
+    Point R = R1.V + R2.V;
+    Point X = kp1.V + kp2.V;
+
+    auto sig1 = sign(kp1.v, X, R, R1.v, secret);
+    auto sig2 = sign(kp2.v, X, R, R2.v, secret);
+    auto sig3 = Signature(R, sig1.s + sig2.s);
+
+    // No one can verify any of those individually
+    assert(!verify(kp1.V, sig1, secret));
+    assert(!verify(kp1.V, sig2, secret));
+    assert(!verify(kp2.V, sig2, secret));
+    assert(!verify(kp2.V, sig1, secret));
+    assert(!verify(kp1.V, sig3, secret));
+    assert(!verify(kp2.V, sig3, secret));
+
+    // But multisig works
+    assert(verify(X, sig3, secret));
+}
+
+/// Batch verification
+unittest
+{
+    enum BatchSize = 42;
+
+    Pair[BatchSize] keys = void;
+    for (size_t i = 0; i < keys.length; ++i)
+        keys[i] = Pair.random();
+
+    Hash[BatchSize] secrets = void;
+    Signature[BatchSize] sigs = void;
+    foreach (idx, ref key; keys)
+    {
+        secrets[idx] = hashFull(idx);
+        sigs[idx] = sign(key, secrets[idx]);
+    }
+
+    assert(verifyBatch(sigs[], keys[].map!(v => v.V), secrets[]));
+}
+
+/// Represent a signature (R, s)
+public struct Signature
+{
+    /// Commitment
+    public Point R;
+    /// Proof
+    public Scalar s;
+}
+
+/// Represent the message to hash (part of `c`)
+private struct Message (T)
+{
+    public Point X;
+    public Point R;
+    public T     message;
+
+    public void computeHash(scope HashDg dg) const nothrow @nogc
```

Comment:
Missing docs.

Reply:
It's a private struct, and a "well-known" hook, so is `///` okay ?